allennlp/commands/diff.py

```diff
@@ -0,0 +1,282 @@
+import torch
+
+from allennlp.commands.subcommand import Subcommand
+from allennlp.common.file_utils import cached_path
+from allennlp.nn.util import load_state_dict
+
+
+logger = logging.getLogger(__name__)
+
+
+@Subcommand.register("diff")
+class Diff(Subcommand):
+    requires_plugins: bool = False
+
+    @overrides
+    def add_subparser(self, parser: argparse._SubParsersAction) -> argparse.ArgumentParser:
+        description = """Display a diff between two model checkpoints."""
+        long_description = (
+            description
+            + """
+        In the output, lines start with either a "+", "-", "!", or empty space " ".
+        "+" means the corresponding parameter is present in the 2nd checkpoint but not the 1st.
+        "-" means the corresponding parameter is present in the 1st checkpoint but not the 2nd.
+        "!" means the corresponding parameter is present in both, but has different weights (same shape)
+        according to the distance calculation and the '--threshold' value.
+        And " " means the corresponding parameter is considered identical in both, i.e.
+        the distance falls below the threshold.
+        The distance between two tensors is calculated as the root of the
+        mean squared difference, multiplied by the '--scale' parameter.
+        """
+        )
+        subparser = parser.add_parser(
+            self.name,
+            description=long_description,
+            help=description,
+        )
+        subparser.set_defaults(func=_diff)
+        subparser.add_argument(
+            "checkpoint1",
+            type=str,
+            help="""the URL, path, or other identifier (see '--checkpoint-type-1')
+            to the 1st PyTorch checkpoint.""",
+        )
+        subparser.add_argument(
+            "checkpoint2",
+            type=str,
+            help="""the URL, path, or other identifier (see '--checkpoint-type-2')
+            to the 2nd PyTorch checkpoint.""",
+        )
+        for i in range(1, 3):
+            subparser.add_argument(
+                f"--checkpoint-type-{i}",
+                type=str,
+                choices=["file", "huggingface"],
+                default="file",
+                help=f"""the type of checkpoint corresponding to the 'checkpoint{i}' argument.
+                If 'file', 'checkpoint{i}' should point directly to a checkpoint file (e.g. '.pt', '.bin')
+                or an AllenNLP model archive ('.tar.gz').
+                If 'huggingface', 'checkpoint{i}' should be the name of a model on HuggingFace's model hub.""",
+            )
+        subparser.add_argument(
+            "--strip-prefix-1",
+            type=str,
+            help="""a prefix to remove from all of the 1st checkpoint's keys.""",
+        )
+        subparser.add_argument(
+            "--strip-prefix-2",
+            type=str,
+            help="""a prefix to remove from all of the 2nd checkpoint's keys.""",
+        )
+        subparser.add_argument(
```

Comment:
I don't completely understand the purpose of the scale arg, could you explain its necessity?

Reply:
We only show a limited number of decimal places (4) in the output, so if you care about differences on the order of 1e-5 or smaller, you may want to increase the `--scale` parameter by an order of magnitude or more. Does that make sense? Should I include that in the help string?